server/src/main/java/org/opensearch/index/store/CompositeDirectory.java

```diff
@@ -0,0 +1,159 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.index.store;
+
+import org.apache.lucene.store.Directory;
+import org.apache.lucene.store.FSDirectory;
+import org.apache.lucene.store.FilterDirectory;
+import org.apache.lucene.store.IOContext;
+import org.apache.lucene.store.IndexInput;
+import org.apache.lucene.store.IndexOutput;
+import org.apache.lucene.store.Lock;
+import org.opensearch.index.store.remote.file.OnDemandCompositeBlockIndexInput;
+import org.opensearch.index.store.remote.filecache.CachedIndexInput;
+import org.opensearch.index.store.remote.filecache.FileCache;
+import org.opensearch.index.store.remote.utils.filetracker.FileState;
+import org.opensearch.index.store.remote.utils.filetracker.FileType;
+
+import java.io.IOException;
+import java.util.Collection;
+import java.util.Set;
+
+public class CompositeDirectory extends FilterDirectory {
+
+    private final FSDirectory localDirectory;
+    private final RemoteStoreFileTrackerAdapter remoteStoreFileTrackerAdapter;
+    private final FileCache fileCache;
+    private final FSDirectory localCacheDirectory;
+
+    public CompositeDirectory(FSDirectory localDirectory, FSDirectory localCacheDirectory, FileCache fileCache) {
+        super(localDirectory);
+        this.localDirectory = localDirectory;
+        this.fileCache = fileCache;
+        this.remoteStoreFileTrackerAdapter = new CompositeDirectoryRemoteStoreFileTrackerAdapter(fileCache);
+        this.localCacheDirectory = localCacheDirectory;
+    }
+
+    public void setRemoteDirectory(Directory remoteDirectory) {
+        ((CompositeDirectoryRemoteStoreFileTrackerAdapter)remoteStoreFileTrackerAdapter).setRemoteDirectory(remoteDirectory);
+    }
+
+    @Override
+    public String[] listAll() throws IOException {
+        return localDirectory.listAll();
+    }
+
+    @Override
+    public void deleteFile(String name) throws IOException {
+        super.deleteFile(name);
+        remoteStoreFileTrackerAdapter.removeFileFromTracker(name);
+        fileCache.remove(localDirectory.getDirectory().resolve(name));
+    }
+
+    @Override
+    public long fileLength(String name) throws IOException {
+        return localDirectory.fileLength(name);
```

Comment:
Will the file be present locally always (maybe ensuring that it has been fetched from the remote?)